tests/ztest/fff_fake_contexts/src/main.c

```diff
@@ -0,0 +1,328 @@
+/*
+ * Copyright (c) 2023 Legrand North America, LLC.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/code_under_test.h>
+
+
+#include <zephyr/ztest.h>
+#include <zephyr/fff.h>
+#include <zephyr/fff_extensions.h>
+
+#include "fakes/called_API.h"
+
+
+#define RESET_HISTORY_AND_FAKES() \
+	ZEPHYR_CALLED_API_FFF_FAKES_LIST(RESET_FAKE) \
+	FFF_RESET_HISTORY()
+
+DEFINE_FFF_GLOBALS;
+
+
+/*
+ * Custom Fakes:
+ */
+
+struct called_API_open_custom_fake_context {
+	/* Written to code under test by custom fake */
+	const struct called_API_info * const instance_out;
+
+	int result;
+};
+
+int called_API_open_custom_fake(
+	const struct called_API_info **instance_out)
+{
+	RETURN_HANDLED_CONTEXT(
+		called_API_open,
+		struct called_API_open_custom_fake_context,
+		result,  /* return field name in _fake_context struct */
+		context, /* Name of context ptr variable used below */
+		{
+			if (context != NULL) {
+				if (context->result == 0) {
+					if (instance_out != NULL) {
+						*instance_out = context->instance_out;
+					}
+				}
+
+				return context->result;
+			}
+
+			return called_API_open_fake.return_val;
+		}
+	);
+}
+
+/*
+ * Tests
+ */
+
+ZTEST(fff_fake_contexts_tests, test_code_under_test)
+{
+	struct test_case {
+		void **expected_call_history;
+
+		/* Last FFF sequence entry is reused for excess calls.
+		 * Have an extra entry that returns a distinct failure (-E2BIG)
+		 *
+		 * Expect one less call than _len, or 0 if sequence ptr is NULL
+		 *
+		 * Configure to return -E2BIG if excess calls.
+		 */
+		int    called_API_open_custom_fake_contexts_len;
+		struct called_API_open_custom_fake_context *
+		       called_API_open_custom_fake_contexts;
+
+		int  called_API_close_fake_return_val_seq_len;
+		int *called_API_close_fake_return_val_seq;
+
+		int result_expected;
+	}
+	const test_cases[] = {
```

Comment:
Please split into separate `ZTEST` functions or add custom printing logic to the asserts below so we know what exactly failed if this ever breaks.

Reply:
Custom print logic added.